File: mobile/utils/polyline.ts
/**
 * Decodes a Google-encoded polyline to latitude/longitude coordinates.
 */
export function decodePolyline(encoded: string): { latitude: number; longitude: number }[] {
  const coords: { latitude: number; longitude: number }[] = [];
  let index = 0;
  let lat = 0;
  let lng = 0;
  while (index < encoded.length) {
    const dlat = decodeValue(encoded, index);
    lat += dlat.value;
    index = dlat.index;
    const dlng = decodeValue(encoded, index);
    lng += dlng.value;
    index = dlng.index;
    coords.push({ latitude: lat * 1e-5, longitude: lng * 1e-5 });
  }
  return coords;
}

function decodeValue(encoded: string, index: number): { value: number; index: number } {
  let result = 0;
  let shift = 0;
  let b: number;
  do {
    b = encoded.charCodeAt(index++) - 63;
    result |= (b & 0x1f) << shift;
    shift += 5;
  } while (b >= 0x20);
  const delta = result & 1 ? ~(result >> 1) : result >> 1;
  return { value: delta, index };
}

File: backend/polyline.py
"""Decode Google encoded polylines to (lat, lng) tuples."""


def decode(encoded: str) -> list[tuple[float, float]]:
    index = 0
    lat = 0
    lng = 0
    coords: list[tuple[float, float]] = []
    while index < len(encoded):
        lat_change, index = _decode_value(encoded, index)
        lng_change, index = _decode_value(encoded, index)
        lat += lat_change
        lng += lng_change
        coords.append((lat * 1e-5, lng * 1e-5))
    return coords


def _decode_value(encoded: str, index: int) -> tuple[int, int]:
    result = 0
    shift = 0
    while True:
        b = ord(encoded[index]) - 63
        index += 1
        result |= (b & 0x1F) << shift
        shift += 5
        if b < 0x20:
            break
    delta = ~(result >> 1) if (result & 1) else (result >> 1)
    return delta, index
